gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -193,76 +198,87 @@ public Object call() throws IOException {
-        uploadId = startResumableUpload();
-        InsertChunkResponseObserver responseObserver;
-
-        long writeOffset = retriesAttempted > 1 ? getCommittedWriteSize(uploadId) : 0;
-
-        responseObserver = new InsertChunkResponseObserver(uploadId, writeOffset);
-        // TODO(b/151184800): Implement per-message timeout, in addition to stream timeout.
-        StreamObserver<InsertObjectRequest> requestStreamObserver =
-            stub.withDeadlineAfter(WRITE_STREAM_TIMEOUT.toMillis(), MILLISECONDS)
-                .insertObject(responseObserver);
-
-        // Wait for streaming RPC to become ready for upload.
-        try {
-          responseObserver.ready.await();
-        } catch (InterruptedException e) {
-          Thread.currentThread().interrupt();
-          throw new IOException(
-              String.format(
-                  "Streaming RPC failed to become ready for resumable upload for '%s'",
-                  getResourceString()),
-              e);
-        }
-        boolean objectFinalized = false;
-        while (!objectFinalized) {
-          InsertObjectRequest insertRequest;
-          if (dataChunkMap.size() > 0 && dataChunkMap.lastKey() >= writeOffset) {
-            insertRequest = buildRequestFromBufferedDataChunk(dataChunkMap, writeOffset);
-            writeOffset += insertRequest.getChecksummedData().getContent().size();
-          } else {
-            ByteString data =
-                ByteString.readFrom(ByteStreams.limit(pipeSource, MAX_BYTES_PER_MESSAGE));
-            dataChunkMap.put(writeOffset, data);
-            if (dataChunkMap.size() >= NUMBER_OF_REQUESTS_TO_RETAIN) {
-              dataChunkMap.remove(dataChunkMap.firstKey());
-            }
-            insertRequest = buildInsertRequest(writeOffset, data, false);
-            writeOffset += data.size();
-          }
-          requestStreamObserver.onNext(insertRequest);
-          objectFinalized = insertRequest.getFinishWrite();
-
-          if (responseObserver.hasTransientError() || responseObserver.hasNonTransientError()) {
-            requestStreamObserver.onError(
-                responseObserver.hasTransientError()
-                    ? responseObserver.transientError
-                    : responseObserver.nonTransientError);
-            break;
-          } else if (objectFinalized) {
-            requestStreamObserver.onCompleted();
-          }
-        }
-
-        try {
-          responseObserver.done.await();
-          if (responseObserver.hasTransientError()) {
-            continue;
-          }
-        } catch (InterruptedException e) {
-          Thread.currentThread().interrupt();
-          throw new IOException(
-              String.format("Resumable upload failed for '%s'", getResourceString()), e);
-        }
+      Retryer<Boolean> retryer =
+          RetryerBuilder.<Boolean>newBuilder()
+              .retryIfExceptionOfType(IOException.class)
+              .retryIfRuntimeException()
```

Comment:
Good point, the runtime exp comes from CounddownLatch thread interruption. Am not sure if this is a good case to retry tho,  maybe let's remove the runtime retry until we fully figure out when/if it can happen and what the impact is.